### PR TITLE
Generated CLI Flags

### PR DESCRIPTION
**Pull request by matkam**
Uses Go `reflect` to generate hidden CLI flags for *most* `Config` struct fields.

```
   --audio.active_level value                               generated (default: 0) [$LIVEKIT_AUDIO_ACTIVE_LEVEL]
   --audio.min_percentile value                             generated (default: 0) [$LIVEKIT_AUDIO_MIN_PERCENTILE]
   --audio.smooth_intervals value                           generated (default: 0) [$LIVEKIT_AUDIO_SMOOTH_INTERVALS]
   --audio.update_interval value                            generated (default: 0) [$LIVEKIT_AUDIO_UPDATE_INTERVAL]
   --development                                            generated (default: false)
   --ingress.rtmp_base_url value                            generated [$LIVEKIT_INGRESS_RTMP_BASE_URL]
   --key_file value                                         generated [$LIVEKIT_KEY_FILE]
   --limit.bytes_per_sec value                              generated (default: 0) [$LIVEKIT_LIMIT_BYTES_PER_SEC]
   --limit.num_tracks value                                 generated (default: 0) [$LIVEKIT_LIMIT_NUM_TRACKS]
   --log_level value                                        generated [$LIVEKIT_LOG_LEVEL]
   --logging.pion_level value                               generated [$LIVEKIT_LOGGING_PION_LEVEL]
   --node_selector.cpu_load_limit value                     generated (default: 0) [$LIVEKIT_NODE_SELECTOR_CPU_LOAD_LIMIT]
   --node_selector.kind value                               generated [$LIVEKIT_NODE_SELECTOR_KIND]
   --node_selector.sort_by value                            generated [$LIVEKIT_NODE_SELECTOR_SORT_BY]
   --node_selector.sysload_limit value                      generated (default: 0) [$LIVEKIT_NODE_SELECTOR_SYSLOAD_LIMIT]
   --port value                                             generated (default: 0) [$LIVEKIT_PORT]
   --prometheus_port value                                  generated (default: 0) [$LIVEKIT_PROMETHEUS_PORT]
   --redis.address value                                    generated [$LIVEKIT_REDIS_ADDRESS]
   --redis.db value                                         generated (default: 0) [$LIVEKIT_REDIS_DB]
   --redis.password value                                   generated [$LIVEKIT_REDIS_PASSWORD]
   --redis.sentinel_master_name value                       generated [$LIVEKIT_REDIS_SENTINEL_MASTER_NAME]
   --redis.sentinel_password value                          generated [$LIVEKIT_REDIS_SENTINEL_PASSWORD]
   --redis.sentinel_username value                          generated [$LIVEKIT_REDIS_SENTINEL_USERNAME]
   --redis.use_tls                                          generated (default: false)
   --redis.username value                                   generated [$LIVEKIT_REDIS_USERNAME]
   --room.auto_create                                       generated (default: false)
   --room.empty_timeout value                               generated (default: 0) [$LIVEKIT_ROOM_EMPTY_TIMEOUT]
   --room.enable_remote_unmute                              generated (default: false)
   --room.max_metadata_size value                           generated (default: 0) [$LIVEKIT_ROOM_MAX_METADATA_SIZE]
   --room.max_participants value                            generated (default: 0) [$LIVEKIT_ROOM_MAX_PARTICIPANTS]
   --rtc.allow_tcp_fallback                                 generated (default: false)
   --rtc.congestion_control.allow_pause                     generated (default: false)
   --rtc.congestion_control.enabled                         generated (default: false)
   --rtc.congestion_control.min_channel_capacity value      generated (default: 0) [$LIVEKIT_RTC_CONGESTION_CONTROL_MIN_CHANNEL_CAPACITY]
   --rtc.congestion_control.padding_mode value              generated [$LIVEKIT_RTC_CONGESTION_CONTROL_PADDING_MODE]
   --rtc.congestion_control.send_side_bandwidth_estimation  generated (default: false)
   --rtc.force_tcp                                          generated (default: false)
   --rtc.node_ip value                                      generated [$LIVEKIT_RTC_NODE_IP]
   --rtc.packet_buffer_size value                           generated (default: 0) [$LIVEKIT_RTC_PACKET_BUFFER_SIZE]
   --rtc.pli_throttle.high_quality value                    generated (default: 0) [$LIVEKIT_RTC_PLI_THROTTLE_HIGH_QUALITY]
   --rtc.pli_throttle.low_quality value                     generated (default: 0) [$LIVEKIT_RTC_PLI_THROTTLE_LOW_QUALITY]
   --rtc.pli_throttle.mid_quality value                     generated (default: 0) [$LIVEKIT_RTC_PLI_THROTTLE_MID_QUALITY]
   --rtc.port_range_end value                               generated (default: 0) [$LIVEKIT_RTC_PORT_RANGE_END]
   --rtc.port_range_start value                             generated (default: 0) [$LIVEKIT_RTC_PORT_RANGE_START]
   --rtc.tcp_port value                                     generated (default: 0) [$LIVEKIT_RTC_TCP_PORT]
   --rtc.udp_port value                                     generated (default: 0) [$LIVEKIT_RTC_UDP_PORT]
   --rtc.use_external_ip                                    generated (default: false)
   --rtc.use_ice_lite                                       generated (default: false)
   --turn.cert_file value                                   generated [$LIVEKIT_TURN_CERT_FILE]
   --turn.domain value                                      generated [$LIVEKIT_TURN_DOMAIN]
   --turn.enabled                                           generated (default: false)
   --turn.external_tls                                      generated (default: false)
   --turn.key_file value                                    generated [$LIVEKIT_TURN_KEY_FILE]
   --turn.relay_range_end value                             generated (default: 0) [$LIVEKIT_TURN_RELAY_RANGE_END]
   --turn.relay_range_start value                           generated (default: 0) [$LIVEKIT_TURN_RELAY_RANGE_START]
   --turn.tls_port value                                    generated (default: 0) [$LIVEKIT_TURN_TLS_PORT]
   --turn.udp_port value                                    generated (default: 0) [$LIVEKIT_TURN_UDP_PORT]
   --video.dynacast_pause_delay value                       generated (default: 0) [$LIVEKIT_VIDEO_DYNACAST_PAUSE_DELAY]
   --webhook.api_key value                                  generated [$LIVEKIT_WEBHOOK_API_KEY]
```

Related to https://github.com/livekit/livekit/pull/731

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/livekit/livekit-server
 
-go 1.17
+go 1.18
```
Comment (boks1971) [inline go.mod:3]: Are we okay with this? I have some vague recollection of something failing with this. Maybe, @cnderrauber faced it? Or maybe it was from community.

File: .github/workflows/buildtest.yaml [hunk 2]
```diff
@@ -21,7 +21,11 @@ jobs:
       - name: Set up Go
         uses: actions/setup-go@v3
         with:
-          go-version: '>=1.17'
+          go-version: '>=1.18'
```
Comment (boks1971) [inline .github/workflows/buildtest.yaml:24]: Left a comment below in `go.mod`. Need to check if this version is okay.

File: pkg/config/config.go [hunk 3]
```diff
@@ -206,7 +211,7 @@ type IngressConfig struct {
 	RTMPBaseURL string `yaml:"rtmp_base_url"`
 }
 
-func NewConfig(confString string, strictMode bool, c *cli.Context) (*Config, error) {
+func NewConfig(c *cli.Context, baseFlags []cli.Flag, strictMode bool) (*Config, error) {
```
Comment (boks1971) [inline pkg/config/config.go:214]: Does this work okay (i. e. passing in `cli.Context` and reading `config-body`) if this module is used from `cloud`?

File: pkg/config/config.go [hunk 4]
```diff
@@ -347,7 +358,172 @@ func (conf *Config) IsTURNSEnabled() bool {
...
+			value := currNode.TypeNode.Field(i)
+			if value.Kind() == reflect.Struct {
+				nodes = append(nodes, configNode{value, yamlPath})
+			} else {
+				flagNames[yamlPath] = value
+			}
+		}
+	}
+
+	return flagNames
```
Comment (boks1971) [inline pkg/config/config.go:404]: wow, that is some juggling to get to the bottom

File: go.mod (hunk 1 above)
Comment (matkam) [reply go.mod:3]: We'd lose a generated CLI flag for a config field that's a pointer (`rtc.allow_tcp_fallback`), but maybe that's OK.

Comment (matkam) [reply go.mod]: Correction - `reflect.Ptr` works for backwards compatibility 👍

File: pkg/config/config.go (hunk 3 above)
Comment (matkam) [reply pkg/config/config.go]: good point. I've reverted this change.

File: cmd/server/main_test.go [hunk 5]
```diff
@@ -43,7 +43,7 @@ func TestShouldReturnErrorIfConfigFileDoesNotExist(t *testing.T) {
 func writeConfigFile(test testStruct, t *testing.T) {
 	if test.configFileName != "" {
 		d1 := []byte(test.expectedConfigBody)
-		err := os.WriteFile(test.configFileName, d1, 0644)
+		err := os.WriteFile(test.configFileName, d1, 0o644)
```
Comment (davidzhao) [inline cmd/server/main_test.go:46]: oh.. TIL!

**Review by davidzhao — APPROVED** [review]
nice work!

File: cmd/server/main_test.go (hunk 5 above)
Comment (matkam) [reply cmd/server/main_test.go:46]: My code formatter did it 🤷‍♂️

**Pull request merged**